### ISSUE TITLE
[@typescript-eslint/quote-props] Extension Rule Proposal

### ISSUE THREAD

**Issue by mpsijm** [typescript-eslint/typescript-eslint#3301]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "quote-props": ["error", "as-needed"],
  }
}
```

```TS
const obj1 = {a: 42}; // OK
const obj2 = {"b": 42}; // ESLint: Unnecessarily quoted property 'b' found.(quote-props)
type obj1 = {a: number}; // OK
type obj2 = {"b": number}; // Expected ESLint error here
enum Foo { a = 1 } // OK
enum Foo { "a" = 1 } // Expected ESLint error here
```

**Expected Result**

Expected ESLint to show an error on the unnecessarily quoted property in the object type literal.

**Actual Result**

No error, because the `quote-props` rule has not been extended yet.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.22.0` |
| `@typescript-eslint/parser`        | `4.22.0` |
| `TypeScript`                       | `4.1.5` |
| `ESLint`                           | `7.24.0` |
| `node`                             | `14.16.1` |

**Comment by bradzacher** [comment]
We generally recommend that you use [`prettier`](https://www.npmjs.com/package/prettier) instead of relying on stylistic lint rules.

**Comment by jpike88** [comment]
I actually use prettier, but only for formatting. Bringing the whole linting side of it in just for this rule would cause our big projects to flip out, it's not an option for me

**Comment by jpike88** [comment]
bonus points if the properties are auto-corrected on save

**Comment by bradzacher** [comment]
prettier actually handles most of the cases.

For some reason it doesn't convert class props or enum members though.

https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEEBGArARkwF5NgBDJTAFgCYBfAbgB0pIoNt9aSzmQc-KnSatWYADbk0aTADEIEMpnK8i9cVJnzFy-uX5rMGqKxgBPAA5wuhXhSpQArgFsccAE6jYVm7jw8pMD8gsiYzm6e3qwAlrCeAGbkYH74RA7hru5esfEeSSm2PMECQpmROaZQCK46SsAqRiY1LnV6IAYgTSAANCAQljAx6Mig5B4eEADuAArjCGjIHRJT5OaLfTgeyQDWcDAAypbJcQDmyDAeTnB9cJEAJvdw9wAy5FCnTuSncAoeLuQYEMPktyE4YBBeiAABYwFwSADq0Ji8DQxxSBwWKJiADcUeYlmAZFC4mhPDAZttTgDkEkJGS+ng0AAPABC2zAe0O5BccBecTgtPI9JuICZzIOZwkcAAik4IPAhSK+scPGSPEsLNY0GAPDFBlDLHrYAiYvcYNDkAAOAAMKsmZIR20sSyNcHVOMFfQAjvL4JSBosUNIALTVZ7PKEeOC+mLRynfGlIOkMkBklwxC5XUVoKWyv2C5PC1Mwcg4U3my1IWh9S7kGISM4AYQgLiTIHdAFYoU4yQAVMtBlOinHXACSUCesAOuv1MAAgpODhZpUqyfR6EA

**Comment by jpike88** [comment]
```
TagGroupType['my-plant']
```

Just got caught out HARD because I didn't find all instances of enum references in code base containing symbols. So they pointed to nothing.

The only (ugly) solution I can come up with now that actually works is a regex check using 'eslint-plugin-regex':

```
[A-z]\['([\w]+?)-
```

**Comment by Josh-Cena** [comment]
Because the ESLint core is also working on [deprecating/removing its formatting rules](https://github.com/eslint/eslint/issues/17522), we will close this. A 3rd-party plugin providing formatting, like https://github.com/eslint-stylistic/eslint-stylistic, can surely support TS OOTB.
